### PR TITLE
Improve observation manager for observation groups with `history_length != None`. 

### LINKED ISSUE #1957: [Bug Report] Reshape + shape computation issues with observation manager
#### Description
The observation manager has a couple issues when providing `history_length != None`. In particular:
1. The `env.observation_shape` is computed incorrectly for these groups when `flatten_history_dim=False`. (See below for a concrete example).
2. When `flatten_history_dim=True` the observation is not returned in a way that is easy to reshape to recover the common history dimension. 

While 2 may be a feature request (I thought this was the expected behavior, but realized it is unclear in hindsight), 1 is certainly a bug with how the observations are being handled currently. 

#### Steps to Reproduce
1. Create an observation group with multiple history-enabled terms. For example:
```
@configclass
class ObservationManagerCfg:
    @configclass
    class PolicyCfg(ObservationGroupCfg):
        history_length = H
        flatten_history_dim = False
        term1 = ObservationTermCfg(func=f1) # f1 returns shape (D1,)
        term2 = ObservationTermCfg(func=f2) # f2 returns shape (D2,)

    policy = PolicyCfg()
```
*Expected:*  Upon instantiating the env, `env.observation_space["policy"].shape` is  `(num_envs, H, D1 + D2)`. 
*Actual*: Upon instantiating the env, `env.observation_space["policy"].shape` is  `(num_envs, 2H, D1 + D2)`. 

2. Create an observation group with `flatten_history_dim=True` (e.g., change this parameter of the last example.

When reading an observation from the observation manager,
```
policy_obs = observation_manager.compute()["policy"]
```
*Expected:* `policy_obs.reshape(-1,  H, D1 + D2) = torch.concatenate([term1, term2], dim=-1)`. Note this is poor syntax - but ideally we could reconstruct the observations with a coherent history dim.
*Actual:* `policy_obs = torch.concatenate([term1.reshape(num_envs, -1), term2.reshape(num_envs, -1)], dim=-1).` The reshape does not preserve a history dimension.

### System Info

Describe the characteristic of your environment:
- Commit: f1a4975eb7bae8509082a8ff02fd775810a73531
- Isaac Sim Version: 4.5.0
- OS: Ubuntu 22.04
- GPU: RTX 5000 Ada
- CUDA: 12.2
- GPU Driver: 535.183.01

### Additional context

Will open a PR with suggested changes to fix these issues.

### Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have checked that the issue is not in running Isaac Sim itself and is related to the repo

### Acceptance Criteria

- [ ] Concatenated, unflattened histories have the correct shape in `env.observation_space`
- [ ] History observations can be reshaped easily to preserve a history dim.

### PR DESCRIPTION
**Pull request by pculbertson-bdai**
# Description
This PR makes some low-level changes to the `observation_manager`, particularly w/r/t how it handles `ObservationGroups` with non-zero `history_length`. It first changes how the observation group shapes are computed - which fixes a shape bug in the `env.observation_space` for this group downstream.

It also changes the handling of flattened obs groups to construct tensors in a way that the history dimension is preserved. 

Fixes #1957. 

## Type of change

Unsure if this is a bugfix or breaking change, since it would change the ordering of elements in observation groups with `history_length != None`. Policies trained with the old ordering would not work - but the proposed ordering seems more in line with the shapes described in the documentation. 

- Bug fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 1]
```diff
@@ -492,3 +506,69 @@ def _prepare_terms(self):
...
+        # Separate flattened and unflattened dimensions.
+        flattened = [dims for dims in group_term_dims if len(dims) == 1]
+        unflattened = [dims for dims in group_term_dims if len(dims) > 1]
+
+        if flattened and unflattened:
+            if self._group_obs_concatenate[group_name]:
+                # Observation shapes should not be mixed if concatenating - raise error.
+                raise RuntimeError(f"In group '{group_name}', mixed dimension formats encountered: {group_term_dims}")
+            else:
+                return group_term_dims
```
Comment (jtigue-bdai) [inline source/isaaclab/isaaclab/managers/observation_manager.py:534]: I think we want to move "if not concatenate return group_term_dims" out of `if flattened and unflattened:` and have it run first.

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 2]
```diff
@@ -492,3 +506,69 @@ def _prepare_terms(self):
...
+
+        if unflattened:
+            # Unflattened: dims = (H, d1, d2, ..., d_n)
+            histories = [dims[0] for dims in unflattened]
+            if not all(h == histories[0] for h in histories):
+                raise RuntimeError(
+                    f"In group '{group_name}', not all observation terms have the same history dimension:"
+                    f" {group_term_dims}"
+                )
```
Comment (jtigue-bdai) [inline source/isaaclab/isaaclab/managers/observation_manager.py:543]:
```suggestion
            if not all(h == histories[0] for h in histories):
                raise RuntimeError(
                    f"In observation group '{group_name}', not all observation terms have the same history dimension:"
                    f" {group_term_dims}. These terms cannot be concatenated if not flattened."
                )
```

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 3]
```diff
@@ -492,3 +506,69 @@ def _prepare_terms(self):
...
+            common_history = histories[0]
+
+            trailing_dims = [dims[1:] for dims in unflattened]
+            trailing_rank = len(trailing_dims[0])
+            if not all(len(td) == trailing_rank for td in trailing_dims):
+                raise RuntimeError(
+                    f"In group '{group_name}', observation terms have different trailing ranks: {group_term_dims}"
+                )
```
Comment (jtigue-bdai) [inline source/isaaclab/isaaclab/managers/observation_manager.py:551]:
```suggestion
            if not all(len(td) == trailing_rank for td in trailing_dims):
                raise RuntimeError(
                    f"In observation group '{group_name}', observation terms have different trailing ranks: {group_term_dims}. These terms cannot be concatenated if not flattened."
                )
```

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 4]
```diff
@@ -492,3 +506,69 @@ def _prepare_terms(self):
...
+
+            if trailing_rank == 0:
+                combined_trailing = ()
+            else:
+                # All trailing dims except the last must match.
+                base = trailing_dims[0][:-1]
+                if not all(td[:-1] == base for td in trailing_dims):
+                    raise RuntimeError(
+                        f"In group '{group_name}', trailing dimensions except last must be the same: {group_term_dims}"
+                    )
```
Comment (jtigue-bdai) [inline source/isaaclab/isaaclab/managers/observation_manager.py:561]:
```suggestion
                if not all(td[:-1] == base for td in trailing_dims):
                    raise RuntimeError(
                        f"In observation group '{group_name}', trailing dimensions except last must be the same: {group_term_dims}. These terms cannot be concatenated if not flattened."
                    )
```

File: source/isaaclab/test/managers/test_observation_manager.py [hunk 5]
```diff
@@ -475,24 +543,46 @@ class PolicyCfg(ObservationGroupCfg):
...
+
+        expected_obs_data_t0 = torch.concat((expected_obs_term_1_data, expected_obs_term_2_data), dim=-1).reshape(
+            self.env.num_envs, -1
+        )
+
         self.assertTrue(torch.equal(expected_obs_data_t0, obs_policy))
+
         # test that the history buffer holds previous data
-        for _ in range(GROUP_HISTORY_LENGTH):
+        for _ in range(GROUP_HISTORY_LENGTH - 1):
```
Comment (jtigue-bdai) [inline source/isaaclab/test/managers/test_observation_manager.py:563]: I believe we want to keep the GROUP_HISTORY_LENGTH so that we fill the buffer and take another step to make sure the history buffer handles filling and rolling.

File: source/isaaclab/test/managers/test_observation_manager.py [hunk 6]
```diff
@@ -475,24 +543,46 @@ class PolicyCfg(ObservationGroupCfg):
...
             observations = self.obs_man.compute()
             obs_policy = observations["policy"]
-        expected_obs_term_1_data = torch.ones(self.env.num_envs, 4 * GROUP_HISTORY_LENGTH, device=self.env.device)
-        expected_obs_term_2_data = lin_vel_w_data(self.env).repeat(1, GROUP_HISTORY_LENGTH)
-        expected_obs_data_t10 = torch.concat((expected_obs_term_1_data, expected_obs_term_2_data), dim=-1)
+
+        # compute the increments that the freerange obs manager would add
+        # over the full history.
+        increments = torch.arange(GROUP_HISTORY_LENGTH, device=self.env.device).reshape(1, -1, 1)
```
Comment (jtigue-bdai) [inline source/isaaclab/test/managers/test_observation_manager.py:569]: would have to increase the increments to handle the fill + 1.

**Review by jtigue-bdai — COMMENTED** [review]
Thanks @pculbertson-bdai. In general the fix and feature look good. I have a few comment/questions.

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 7]
```diff
@@ -384,7 +387,12 @@ def _prepare_terms(self):
             # iterate over all the terms in each group
             for term_name, term_cfg in group_cfg_items:
                 # skip non-obs settings
-                if term_name in ["enable_corruption", "concatenate_terms", "history_length", "flatten_history_dim"]:
+                if term_name in [
+                    "enable_corruption",
+                    "concatenate_terms",
+                    "history_length",
+                    "flatten_history_dim",
+                ]:
```
Comment (ooctipus) [inline source/isaaclab/isaaclab/managers/observation_manager.py:395]: I will not make this change

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 8]
```diff
@@ -412,7 +420,9 @@ def _prepare_terms(self):
                 # create history buffers and calculate history term dimensions
                 if term_cfg.history_length > 0:
                     group_entry_history_buffer[term_name] = CircularBuffer(
-                        max_len=term_cfg.history_length, batch_size=self._env.num_envs, device=self._env.device
+                        max_len=term_cfg.history_length,
+                        batch_size=self._env.num_envs,
+                        device=self._env.device,
```
Comment (ooctipus) [inline source/isaaclab/isaaclab/managers/observation_manager.py:425]: I will not make this change

File: source/isaaclab/isaaclab/managers/observation_manager.py [hunk 9]
```diff
@@ -452,7 +462,11 @@ def _prepare_terms(self):
                                         f"Modifier function '{mod_cfg.func}' for observation term '{term_name}'"
                                         f" is not a subclass of 'ModifierBase'. Received: '{type(mod_cfg.func)}'."
                                     )
-                                mod_cfg.func = mod_cfg.func(cfg=mod_cfg, data_dim=obs_dims, device=self._env.device)
+                                mod_cfg.func = mod_cfg.func(
+                                    cfg=mod_cfg,
+                                    data_dim=obs_dims,
+                                    device=self._env.device,
+                                )
```
Comment (ooctipus) [inline source/isaaclab/isaaclab/managers/observation_manager.py:469]: I will not make this change

File: source/isaaclab/isaaclab/managers/observation_manager.py (hunk 2 above)
Comment (ooctipus) [reply source/isaaclab/isaaclab/managers/observation_manager.py:543]: I am a bit confused about here:
original comment makes sense, assuming all unflattened term is caused by history dimension, it is checking the if the first dimension is the same. 

the new suggestion I guess is assuming the unflattened term is not just limited to history dimension, maybe the term itself is returned as high dim, so we tell user you must flatten it. 

Then the doc string above seems to suggest unflattened term is cause by history, wrote: 

```
For unflattened terms (tuples of length >= 2), each dims is expected to be of the form:
            (H, d1, d2, ..., d_n)
```

we need to be more clear what is really happening there, it took me a long time to figure out where the comment is coming frome

File: source/isaaclab/isaaclab/managers/observation_manager.py (hunk 3 above)
Comment (ooctipus) [reply source/isaaclab/isaaclab/managers/observation_manager.py:551]: I don't know if assuming same trailing rank, or telling user if trailing rank doesn't match you must flatten is right move

say if you have a group that has two image terms, say
if history length == 5, image 1 has shape (5, 3, 32, 32) and image 2 has shape (5, 3, 32, 32):
-> we should be able to concatenate to   (5, 6, 32, 32)


if history length == 5, image 1 has shape (5, 32, 32, 3) and image 2 has shape (5, 32, 32, 3):
-> we should be able to concatenate to   (5, 32, 32, 6)


if history length == 5, image 1 has shape (5, 3, 32, 32) and image 2 has shape (5, 3, 32, 32):
-> we should tell user that term shape must match instead of just say flatten it

Comment (ooctipus) [reply source/isaaclab/isaaclab/managers/observation_manager.py:551]: the code is currently very difficult to parse I'd suggest put a example in the comment what this block checking, and if passed this block, what fact we can assume about the this unflattened group data

File: source/isaaclab/isaaclab/managers/observation_manager.py (hunk 4 above)
Comment (ooctipus) [reply source/isaaclab/isaaclab/managers/observation_manager.py:561]: I don't know if assuming concatenate trailing rank, or telling user if trailing rank doesn't match you must flatten is right move

say if you have a group that has two image terms, say
if history length == 5, image 1 has shape (5, 3, 32, 32) and image 2 has shape (5, 3, 32, 32):
-> we should be able to concatenate to (5, 6, 32, 32)

if history length == 5, image 1 has shape (5, 1, 32, 32) and image 2 has shape (5, 3, 32, 32):
-> we should be able to concatenate to (5, 4, 32, 32)

if history length == 5, image 1 has shape (5, 32, 32, 1) and image 2 has shape (5, 32, 32, 3):
-> we should be able to concatenate to (5, 32, 32, 4)

if history length == 5, image 1 has shape (5, 3, 32, 32) and image 2 has shape (5, 32, 32, 1):
-> we should Raise Error tell user that all term's shape must match except the dim to concatenate

**Review by ooctipus — CHANGES_REQUESTED** [review]
Thanks for the PR,
I agree how we are currently handles history + history concatenation is not the most helpful one, it lacks a lot of feature, and is too aggressive when comes to concatenation

This PR is making the right move to promising direction, but the concept of history + preserves dimension vs how to flatten is a very complex topic and we want to be careful and clear about each case and what move to make, ideally if the program has resolve the plan we log to user what concatenation plan has we resolve.

**Comment by AntoineRichard** [comment]
Hi @pculbertson-bdai — thank you so much for opening this PR, and for taking the time to contribute to Isaac Lab! 🙏

We're currently doing a big spring-clean of our pull request backlog, which had grown past 400 open PRs. Being honest: at that volume we stopped being able to give every contribution the attention it deserved, and this one is a case where we dropped the ball on getting back to you. That's on us, not on you.

**Why this PR is being closed:** Here is exactly what we found on this PR when we reviewed the backlog:

| | |
|---|---|
| **Opened** | 2025-02-27 (about a year ago) |
| **Last commit on the branch** | 2025-02-27 |
| **Last activity from the author** | about a year ago |
| **Target branch** | `main` — note that active development happens on `develop`; `main` only receives releases |
| **Review status** | Changes were requested (jtigue-bdai, ooctipus), but the follow-up never landed |
| **Merge status** | Conflicts with the base branch — it no longer applies cleanly |
| **Size** | 4 commit(s), 2 file(s) changed, +226 / -56 |

It was picked up by the sweep because it targets `main` instead of `develop`, and it has been open for about a year. It was then put in the "close" bucket because the author has been silent for about a year — which is the signal we used to tell apart pull requests that are still being worked on from ones that have genuinely been set aside.

We deliberately did **not** close pull requests that were approved and ready to land, or that were small and clearly still fixing a live bug — there were 27 of those, and we are merging them rather than closing them.

To be completely clear — **this is not a judgement on the quality or the value of your work.** It's purely about getting the queue down to a size where we can actually review things properly and give contributors real, timely feedback.

**One thing worth knowing first:** Isaac Lab 3.0 has since shipped. `develop` is now at
3.0.0 while `main` is still at 2.3.2, and 3.0 restructured a fair amount — quaternion
conventions, actuator collections, data access, and the Isaac Sim extension imports all
changed. The [migration guide](https://github.com/isaac-sim/IsaacLab/blob/develop/docs/source/migration/migrating_to_isaaclab_3-0.rst)
is the place to start if you rebase.

It is also the honest reason a lot of these older pull requests stopped applying: the code
they touched moved or was rewritten underneath them.

**If this change is still relevant, please reopen this PR or open a fresh one against `develop`.** With the backlog down to a reviewable size, it should get a response a lot sooner than this one did.

Thanks again for contributing to Isaac Lab, and sorry for the long silence! 💚

---
🤖 *This comment was drafted with AI assistance as part of a maintainer-led sweep of the Isaac Lab pull request backlog. A maintainer is behind this cleanup — but if this closure looks wrong, it may well be, so please push back and we'll take another look.*
